src/main/webapp/app/guided-tour/guided-tour.service.ts

```diff
@@ -0,0 +1,367 @@
+                        } else {
+                            this.backStep();
+                        }
+                    });
+                } else {
+                    if (this._checkSelectorValidity()) {
+                        this._guidedTourCurrentStepSubject.next(this.getPreparedTourStep(this._currentTourStepIndex));
+                    } else {
+                        this.backStep();
+                    }
+                }
+            } else {
+                this.resetTour();
+            }
+        }
+    }
+
+    /**
+     * Skip tour
+     */
+    public skipTour(): void {
+        if (this._currentTour) {
+            if (this._currentTour.skipCallback) {
+                this._currentTour.skipCallback(this._currentTourStepIndex);
+            }
+            this.resetTour();
+        }
+    }
+
+    /**
+     * Close tour and remove overlay
+     */
+    public resetTour(): void {
+        document.body.classList.remove('tour-open');
+        this._currentTour = undefined;
+        this._currentTourStepIndex = 0;
+        this._guidedTourCurrentStepSubject.next(undefined);
+    }
+
+    /**
+     * Start guided tour for given guided tour
+     * @param tour  guided tour
+     */
+    public startTour(tour: GuidedTour): void {
+        this.currentTourSteps = tour.steps;
+
+        // adjust tour steps according to permissions
+        tour.steps.forEach((step, index) => {
+            if (step.permission && !this.accountService.hasAnyAuthorityDirect(step.permission)) {
+                this.currentTourSteps.splice(index, 1);
+            }
+        });
+
+        this._currentTour = cloneDeep(tour);
+        this._currentTour.steps = this._currentTour.steps.filter(step => !step.skipStep);
+        this._currentTourStepIndex = 0;
+        this._setFirstAndLast();
+        this._guidedTourOrbShowingSubject.next(this._currentTour.useOrb);
+        if (this._currentTour.steps.length > 0 && (!this._currentTour.minimumScreenSize || window.innerWidth >= this._currentTour.minimumScreenSize)) {
+            if (!this._currentTour.useOrb) {
+                document.body.classList.add('tour-open');
+            }
+            const currentStep = this._currentTour.steps[this._currentTourStepIndex];
+            if (currentStep.action) {
+                currentStep.action();
+            }
+
+            if (this._checkSelectorValidity()) {
+                this._guidedTourCurrentStepSubject.next(this.getPreparedTourStep(this._currentTourStepIndex));
+            } else {
+                this.nextStep();
+            }
+        }
+    }
+
+    /* Start tour with orb */
+    public activateOrb(): void {
+        this._guidedTourOrbShowingSubject.next(false);
+        document.body.classList.add('tour-open');
+    }
+
+    /**
+     * Define first and last tour step based on amount of tour steps
+     * @private
+     */
+    private _setFirstAndLast(): void {
+        if (this._currentTour) {
```

Comment:
This could also just be a getter of the class, the calculation is very easy and more attributes make the class more complicated.